### PR TITLE
feat(clone): new parameter to identify deleted snapshot having active clones

### PR DESCRIPTION
**Pull request by hrudaya21**

### CODE REVIEW

File: protobuf/v1/snapshot.proto [hunk 1]
```diff
@@ -33,6 +33,7 @@ message SnapshotInfo {
     bool valid_snapshot = 12; // true: valid, false: invalid (missing one/more snapshotdescriptor field i.e. txn_id, entity_id, source_uuid
     bool ready_as_source = 13; // ready for usage, i,e source for another volume.
     uint64 referenced_bytes = 14; // amount of bytes referenced by all successor snapshots.
+    bool future_delete_snapshot = 15; // true: snapshot is deleted and unused. false: Valid Snapshot
```
Comment (tiagolobocastro) [inline protobuf/v1/snapshot.proto]: Can we add a filter to selectively exclude these from the list call?

Comment (hrudaya21) [reply protobuf/v1/snapshot.proto]: > Can we add a filter to selectively exclude these from the list call?

You mean to have a filter flag in ListSnapshot gRPC request parameter to filter these "future delete" snapshot in response ?

Comment (tiagolobocastro) [reply protobuf/v1/snapshot.proto]: Yes, as for most uses control-plane won't need to list these.
Also might be useful similar thing for list Replicas, maybe we don't always need to include snapshots in list_replicas, and list only regular replicas and cloned replicas.

Comment (hrudaya21) [reply protobuf/v1/snapshot.proto]: Added the optional parameters with multiple values.

File: protobuf/v1/replica.proto [hunk 2]
```diff
@@ -87,4 +87,9 @@ message ListReplicaOptions {
   google.protobuf.StringValue poolname = 2; // list the replicas on the provided pool, by name
   google.protobuf.StringValue uuid = 3; // list the replica with the uuid if provided
   google.protobuf.StringValue pooluuid = 4; // list the replicas on the provided pool, by uuid
+  // 0: All Replicas including snapshotclones and snapshots (Default)
+  // 1: Replica and Clones, without snapshots
+  // 2: Only Clones
+  // 3: Only Replicas
+  string replica_type = 5; // list replicas based on replicatype value
```
Comment (tiagolobocastro) [inline protobuf/v1/replica.proto]: This should be enum rather than string?

File: protobuf/v1/snapshot.proto [hunk 3]
```diff
@@ -33,11 +33,18 @@ message SnapshotInfo {
...
 /// Request parameters to list snapshot.
 message ListSnapshotsRequest {
     optional string source_uuid = 1;  // source uuid from which snapshot is taken
     optional string snapshot_uuid = 2; // snapshot uuid for which the detail can be listed
+    // 0: All snapshots (Default).
+    // 1: All snapshot except "futuredelete" snapshot
+    // 2: Only "futuredelete" snapshot.
+    // 3: Only "invalid" snapshot
+    // 4: Only usuable snapshot except "futuredelete" and "invalid" snapshots.
+    optional string snapshot_type = 3; // list snapshot based on snapshottype value
```
Comment (tiagolobocastro) [inline protobuf/v1/snapshot.proto]: Same comment here about enum

File: protobuf/v1/replica.proto (hunk 2 above)
Comment (hrudaya21) [reply protobuf/v1/replica.proto]: Yes. fixed.

File: protobuf/v1/snapshot.proto (hunk 3 above)
Comment (hrudaya21) [reply protobuf/v1/snapshot.proto]: fixed

**Review by tiagolobocastro — APPROVED** [review]

**Review by mtzaurus — APPROVED** [review]

**Pull request merged**